### PR TITLE
Support v1beta2 licenses

### PR DESCRIPTION
**Pull request by crdant**
# Adds support for v1beta1 and v1beta2 license formats

## TL;DR

Implements comprehensive multi-version license support through LicenseWrapper abstraction, enabling customers to use either v1beta1 or v1beta2 license formats interchangeably across all installation, upgrade, and template rendering operations while maintaining full backwards compatibility.

Depends on replicatedhq/kots#5675

## Details

The embedded-cluster codebase previously coupled tightly to the kotsv1beta1.License type, preventing support for newer v1beta2 license formats that include enhanced entitlement structures and metadata. This migration introduces the LicenseWrapper abstraction from kotskinds (commit 174e89c93554), which provides version-agnostic access to license fields and entitlements, allowing the system to transparently handle both license versions without code duplication or conditional logic scattered throughout the codebase.

### Architecture Overview

The implementation follows a three-layer approach:

1. **Parsing Layer** - Updated `pkg/helpers/parse.go` to return `licensewrapper.LicenseWrapper` instead of raw `kotsv1beta1.License` types, with `LoadLicenseFromBytes()` handling version detection and appropriate struct initialization.

2. **Business Logic Layer** - Migrated all production code across CLI commands (`cmd/installer/cli/`), infrastructure managers (`api/internal/managers/`), and template engine (`api/pkg/template/`) to use LicenseWrapper types exclusively, replacing direct struct field access with wrapper methods.

3. **API Layer** - Replaced direct `.Spec.*` field access patterns with wrapper getter methods like `GetAppSlug()`, `GetLicenseID()`, `GetChannelName()`, and `IsEmbeddedClusterMultiNodeEnabled()`, ensuring consistent behavior regardless of underlying license version.

### Key Changes by Component

**Parser Infrastructure (pkg/helpers/)**
- ParseLicense now returns `licensewrapper.LicenseWrapper` with automatic version detection
- ParseLicenseFromBytes delegates to `licensewrapper.LoadLicenseFromBytes()` for both v1beta1 and v1beta2 parsing
- Comprehensive test coverage with fixtures for both versions, invalid versions, and edge cases like missing fields

**Installer CLI (cmd/installer/cli/)**
- Install command updated to use LicenseWrapper throughout configuration, validation, and metrics reporting
- Upgrade command migrated from `*kotsv1beta1.License` to `licensewrapper.LicenseWrapper` in upgradeConfig struct
- License expiration validation refactored to use entitlement abstraction methods instead of direct `.StrVal` access
- All metrics reporters updated to use wrapper methods for license ID and app slug extraction

**Infrastructure Managers (api/internal/managers/)**
- Linux and Kubernetes install managers updated to accept and process LicenseWrapper types
- Upgrade manager parsing refactored to use `LoadLicenseFromBytes()` instead of raw unmarshaling
- LicenseInfo population in Installation CRD updated to call `IsDisasterRecoverySupported()` and `IsEmbeddedClusterMultiNodeEnabled()` wrapper methods
- Resolved critical merge conflict with duplicate license field definition in app config manager

**Template Engine (api/pkg/template/)**
- Engine struct updated to store `licensewrapper.LicenseWrapper` instead of raw license pointer
- All license template functions refactored to use wrapper getter methods for consistent field access
- LicenseFieldValue function updated to use `Value()` method on entitlement fields, supporting both v1beta1 direct values and v1beta2 structured entitlements
- Channel name resolution updated to use `GetChannels()`, `GetChannelID()`, and `GetChannelName()` methods
- Nil safety improved with direct `V1`/`V2` checks instead of method calls where appropriate

**Package-Level Types (pkg/)**
- Metrics reporter updated to accept LicenseWrapper and use getter methods
- Kubeutils installation functions migrated to LicenseWrapper parameters
- Addons installation updated to use wrapper methods for license field access

### Validation and Testing

The implementation includes extensive test coverage:

- TDD approach with test fixtures created before implementation (pkg/helpers/testdata/)
- Tests for v1beta1 license parsing, v1beta2 license parsing, invalid versions, missing required fields
- Integration tests updated with apiVersion/kind metadata for LicenseWrapper compatibility
- CLI tests refactored to use LicenseWrapper constructors and validation methods
- Template engine tests covering both license versions with all template functions
- All production code paths verified to use wrapper methods exclusively

Test results confirm successful migration:
- ✅ Parser tests: PASS (pkg/helpers)
- ✅ Template engine tests: PASS (api/pkg/template)
- ✅ Infrastructure manager tests: PASS (api/internal/managers)
- ✅ Build succeeds: `go build ./...`
- ✅ No direct `.Spec.*` license access in production code
- ✅ No old `*kotsv1beta1.License` types in production code

### Backwards Compatibility

Existing v1beta1 licenses continue to work without any changes:
- LicenseWrapper detects version from apiVersion field and populates appropriate internal struct (V1 or V2)
- Getter methods provide identical return values regardless of version
- License storage remains byte-based with no schema migrations required
- Template syntax unchanged, with all functions working transparently across versions

### Breaking Changes

None. This is a purely internal refactoring that maintains API compatibility:
- CLI flags and arguments unchanged
- Template function signatures identical
- License file formats both supported
- Metrics and reporting behavior preserved

### Impact

Customers can now:
- Use v1beta2 licenses with enhanced entitlement metadata and structured fields
- Continue using existing v1beta1 licenses without modification
- Switch between license versions seamlessly during upgrades
- Access new entitlement features when available in v1beta2 format

The system maintains:
- Full backwards compatibility with all existing v1beta1 licenses
- Consistent behavior across installation, upgrade, and template rendering paths
- Type safety through wrapper methods rather than direct field access
- Extensibility for future license versions through abstraction layer

### Dependency Update

Updated kotskinds to commit 174e89c93554 which includes:
- LicenseWrapper struct with V1 and V2 fields
- Version-agnostic getter methods for all common license fields
- LoadLicenseFromBytes function with automatic version detection
- EntitlementFieldWrapper for abstracted entitlement value access

### Related Documentation

- Implementation Plan: `docs/plans/2025-10-31-complete-licensewrapper-migration.md`
- Research Notes: `docs/research/2025-10-31-license-wrapper-remaining-work.md`
- Test Fixtures: `pkg/helpers/testdata/license-*.yaml`

---

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: pkg/helpers/parse_test.go [hunk 1]
```diff
@@ -113,108 +113,142 @@ kind: Config`,
...
+		wantLicenseID string
+		wantECEnabled bool
+		wantCustomer  string
 	}{
 		{
-			name:    "file does not exist",
-			fpath:   "nonexistent.yaml",
-			wantErr: os.ErrNotExist,
+			name:          "v1beta1 license",
+			licenseFile:   "testdata/license-v1beta1.yaml",
```
Comment (emosbaugh) [inline pkg/helpers/parse_test.go]: can you go embed testdata?

File: cmd/installer/cli/install.go [hunk 2]
```diff
@@ -90,7 +91,7 @@ type installConfig struct {
 	isAirgap                bool
 	enableManagerExperience bool
 	licenseBytes            []byte
-	license                 *kotsv1beta1.License
+	license                 licensewrapper.LicenseWrapper
```
Comment (emosbaugh) [inline cmd/installer/cli/install.go]: it seems safer to pass a pointer to the license wrapper around given that we had a pointer to license. that would allow for `license != nil` rather than `license.GetLicenseID() != ""`

File: api/internal/managers/app/release/manager.go [hunk 3]
```diff
@@ -24,7 +25,7 @@ type AppReleaseManager interface {
 type appReleaseManager struct {
 	rawConfig                  kotsv1beta1.Config
 	releaseData                *release.ReleaseData
-	license                    *kotsv1beta1.License
+	license                    licensewrapper.LicenseWrapper
```
Comment (emosbaugh) [inline api/internal/managers/app/release/manager.go]: and everywhere else make this a pointer

**Review by emosbaugh — CHANGES_REQUESTED** [review]

File: api/pkg/template/license.go [hunk 4]
```diff
@@ -10,65 +10,89 @@ import (
 	"github.com/replicatedhq/embedded-cluster/pkg/release"
 )
 
+// Helper methods for direct access (used by tests and other code)
```
Comment (emosbaugh) [inline api/pkg/template/license.go]: these dont seem to be used in anything other than tests. are they really necessary?

File: api/pkg/template/license_test.go [hunk 5]
```diff
@@ -590,11 +598,121 @@ func TestEngine_ChannelName_ChannelNotFound(t *testing.T) {
...
+	engine := NewEngine(config, WithLicense(wrapLicense(license)), WithReleaseData(releaseData))
 
 	err := engine.Parse("{{repl ChannelName }}")
 	require.NoError(t, err)
 	_, err = engine.Execute(nil, WithProxySpec(&ecv1beta1.ProxySpec{}))
 	assert.Error(t, err)
 	assert.Contains(t, err.Error(), "channel unknown-channel-id not found in license")
 }
+
+func TestEngine_LicenseWrapper(t *testing.T) {
```
Comment (emosbaugh) [inline api/pkg/template/license_test.go]: what does this test?

File: pkg-new/replicatedapi/client.go [hunk 6]
```diff
@@ -144,30 +150,45 @@ func (c *client) newRetryableRequest(ctx context.Context, method string, url str
 
 // injectHeaders injects the basic auth header, user agent header, and reporting info headers into the http.Header.
 func (c *client) injectHeaders(header http.Header) {
-	header.Set("Authorization", "Basic "+basicAuth(c.license.Spec.LicenseID, c.license.Spec.LicenseID))
+	if c.license.IsEmpty() {
```
Comment (emosbaugh) [inline pkg-new/replicatedapi/client.go]: not all of these headers depend on a license

**Review by emosbaugh — CHANGES_REQUESTED** [review]

File: pkg-new/replicatedapi/client.go (hunk 6 above)
Comment (crdant) [reply pkg-new/replicatedapi/client.go]: But don't we need auth?

File: api/pkg/template/license_test.go (hunk 5 above)
Comment (crdant) [reply api/pkg/template/license_test.go]: Right now I think only that it works with a `v1beta2` license, but I can just change one of the other test cases to handle that we we cover everything this test does.

File: api/pkg/template/license.go (hunk 4 above)
Comment (crdant) [reply api/pkg/template/license.go]: I'm reviewing, but you are probably right.
